csrc/scheduler/pointwise_utils.cpp

```diff
@@ -142,6 +142,118 @@ bool DomainMap::areAllInputIdsMappedTo(TensorView* input_tv, TensorView* tv)
+// an easy way to check that.
+// Instead of that, this function checks that all source iter domains involved
+// in transformation on output_tv is covered by reference_tv. We do so by
+// traverse all disjoint set producers on both tvs and filter them with
+// `ca_map_.uniqueExactDefinitions(id).empty()`.
+//
+// ------
+//
+// e.g 0.
+//   T34 [i0, i1]
+//   T185 [i0, b2, i1]     = broadcast(T34)
+//   T192 [i0, b3(ex), i1] = expand(T185)
+//   T198 [i0, b3(ex)*i1]  = reshape(T192)
+//   output(T34)
+//   output(T198)
+//
+// if we consider taking T34 as reference_tv. T198 is the output_tv. We can't
+// replay T34's transform of merging all the dimensions to T198, since b3(ex)*i1
+// can't be reversed. The check in this function would give us T34 with source
+// i0, i1; where T198 would have source i0, b3, i1, where b3 isn't contained in
+// T34. Hence we'll reject this reference_tv.
+//
+// ------
+//
+// e.g 1.
+//   T0 [i0, i1]
+//   T1 [i2, i0, i1]
+//   T2 [i0*i1]      = reshape(T0)
+//   T3 [b3, i0, i1] = broadcast(T0)
+//   T4 [i2, i0, i1] = add(T1, T3)
+//   output(T2)
+//   output(T4)
+//
+// the example above should be able to pick T4 as reference_tv. T2's source i0,
+// i1 are both contained by the source of T4, so this example could be scheduled
+// as a single fusion.
+bool DomainMap::areAllOutputIdsMappedTo(
+    TensorView* output_tv,
+    TensorView* reference_tv) const {
+  // traverse back to collect all disjoint set producers from the logical domain
+  // of tv.
+  auto get_source_producers = [this](TensorView* tv) {
+    VectorOfUniqueEntries<std::shared_ptr<VectorOfUniqueEntries<IterDomain*>>>
+        all_producer_sets;
+    std::for_each(
+        tv->getLogicalDomain().begin(),
+        tv->getLogicalDomain().end(),
+        [&](IterDomain* id) {
+          all_producer_sets.pushBack(
+              ca_map_.disjointSetOf(id, IdMappingMode::EXACT));
+        });
+    all_producer_sets.pushBack(
+        ca_map_.getAllDisjointSetProducers(all_producer_sets));
+
+    std::vector<IterDomain*> source_ids;
+    std::for_each(
+        all_producer_sets.vector().begin(),
+        all_producer_sets.vector().end(),
+        [&source_ids,
+         this](const std::shared_ptr<VectorOfUniqueEntries<IterDomain*>>&
+                   producer_set_ptr) {
+          IterDomain* id = producer_set_ptr->front();
+          if (ca_map_.uniqueExactDefinitions(id).empty()) {
+            source_ids.push_back(id);
+          }
+        });
+    return source_ids;
+  };
+
+  // this contains all source iter domain that's covered by reference_tv, so
+  // it's safe for output_tv to have them.
+  std::unordered_set<IterDomain*> covered_source_ids;
+  for (IterDomain* id : get_source_producers(reference_tv)) {
+    covered_source_ids.insert(id);
+  }
+  // It's safe to have unmapped broadcast IterDomain. There're quite a few tests
+  // expecting pointwise scheduler to handle this pattern
```

Comment:
TL;DR: good call, let me add a test to verify and change it.

Earlier I have this inside the loop below in line 246, I think that wouldn't have the problem.
```
for (IterDomain* id : get_source_producers(output_tv)) {
  if (id->isBroadcast()) {
    continue;
  }
...
}
```

I changed that to this version, because I was concerned that might be too loose a check.
I was mostly worried that if we later decide to add `expand` as a ID op, then we won't be able to identify expand ops. But I think I can add another check along with `ca_map_.uniqueExactDefinitions` for that. At least we have CI guarding that. I'll put a comment there so we'll remember how to fix it then.

Reply:
Logging offline conversation. Due to issue #1126 , we don't see split/merge on broadcast here.
I'm keeping this code as-is. We added an assert that broadcast IDs on logical domain cannot have definition. :crossed_fingers: Let's see if that causes any CI issue.

Reply:
The assertion may not hold when a non-broadcast ID is sliced to a broadcast ID. See https://github.com/NVIDIA/Fuser/pull/3513#issuecomment-2533494215